core: directly load infrastructure from RailJSON

multun opened the pull request:
Fix part of https://github.com/OpenRailAssociation/osrd/issues/7056

First commit is a rebased squash of all PRs previously targeting `core/infra_load_refactor` branch.

(some tests are still using legacy builder)

bougue-pe committed:
core: directly import infrastructure from RailJSON

- removes SimInfraAdapter, which was a leftover of the development of the adapter.
  the old adapter now directly returns a RawInfra
- add a new builder and RawInfra implementation

Co-Authored-By: Victor Collod <victor.collod@epita.fr>

multun committed:
front, editoast: update core parsing progress states

Khoyo commented on core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt:
```diff
@@ -0,0 +1,756 @@
+            }
+        }
+
+        // for each detector pair inside the track section
```
This sentence seems incomplete

Khoyo commented on core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt:
```diff
@@ -0,0 +1,756 @@
+}
+
+/** Build detection zones on a partially-filled builder (track-sections, nodes and detectors) */
+private fun buildZones(builder: RawInfraFromRjsBuilder) {
```
This whole function is quite long, and it's not immediately clear to me what every step does (or rather, why)

multun replied:
I've added steps in an attempt to make it clearer in a separate PR : https://github.com/OpenRailAssociation/osrd/pull/7280

Khoyo commented on core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt:
```diff
@@ -0,0 +1,756 @@
+    for (trackSectionIdx in builder.getTrackSections()) {
+        val detectors = builder.getTrackSectionDetectors(trackSectionIdx)
+
+        // prepare "node" detection zones
```
What are `"node" detection zones`?

multun replied:
the algorithm is correct but weird, I'll refactor it

Khoyo commented on core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt:
```diff
@@ -0,0 +1,756 @@
+    }
+
+    // Parse neutral-sections
+    fun parseNeutralRanges(
```
We seem to suddenly switch to use local function half way through `parseRJSInfra()`...

Khoyo commented on core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt:
```diff
@@ -0,0 +1,756 @@
+                    )
+
+                    // TODO: remove this priority given to announcement.
+                    // Only added to match previous load, but no functional need exists
```
```suggestion
                    // Only added to match the former import process, but no functional need exists
```

Khoyo commented on core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt:
```diff
@@ -0,0 +1,756 @@
+    // parse nodes
+    val switchTypeMap =
+        ((rjsInfra.switchTypes ?: listOf()) + RJSSwitchType.BUILTIN_NODE_TYPES_LIST).associateBy {
+            switchType ->
+            switchType.id
```
if the formatter is going to behave this badly, just use `it.id` ?

Khoyo commented on core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraRJSParser.kt:
```diff
@@ -0,0 +1,756 @@
+    return if (this == EdgeEndpoint.BEGIN) Endpoint.START else Endpoint.END
+}
+
+fun parseRJSInfra(rjsInfra: RJSInfra): RawInfra {
```
Damn, 430 lines long...

We probably should break it up. Not necessarily in this PR, but we should

Khoyo commented on core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/new_impl/RawInfraFromRjsBuilder.kt:
```diff
@@ -0,0 +1,822 @@
+                MutableStaticIdxArrayList(),
+                loadingGaugeConstraints,
+                // Electrifications will be filled later on
+                distanceRangeMapOf(
+                    listOf(DistanceRangeMap.RangeMapEntry(0.meters, length.distance, ""))
+                ),
+                // NeutralSections will be filled later on
+                DirectionalMap(distanceRangeMapOf(), distanceRangeMapOf()),
+                // SpeedSections will be filled later on
+                DirectionalMap(
+                    distanceRangeMapOf(
+                        listOf(
+                            DistanceRangeMap.RangeMapEntry(
+                                0.meters,
+                                length.distance,
+                                SpeedSection(
+                                    Double.POSITIVE_INFINITY.metersPerSecond,
+                                    mapOf(),
+                                    mapOf()
+                                )
+                            )
+                        )
+                    ),
+                    distanceRangeMapOf(
+                        listOf(
+                            DistanceRangeMap.RangeMapEntry(
+                                0.meters,
+                                length.distance,
+                                SpeedSection(
+                                    Double.POSITIVE_INFINITY.metersPerSecond,
+                                    mapOf(),
+                                    mapOf()
+                                )
+                            )
+                        )
+                    )
+                )
+            )
+        )
+    }
```
Huge nit, but this is a 50 lines long statement going ~8 levels of indentation deep

Khoyo reviewed (approved):
Overall LGTM.

Most of my comments are kinda nitpicky, I'll leave to @multun judgement whether to address them now or later.

The process is globally much more readable than the legacy import, but it still feels kinda hard to follow the big picture. I wonder if some restructuring - or more documentation could help.

It's also unclear why some stuff is done in the infra implementation, in the builder and in the parse function - I'm guessing this is still kinda WIP :)

:shipit:

multun committed:
core: move UnionFind to kt-osrd-utils

multun committed:
core: remove the old infrastructure API

This change also removes:
- the old adapter
- the tooling to compare instructures

Co-Authored-By: Pierre-Etienne Bougué <bougue.pe@proton.me>

bougue-pe commented on core/kt-osrd-rjs-parser/src/main/kotlin/fr/sncf/osrd/RawInfraRJSParser.kt:
```diff
@@ -0,0 +1,788 @@
+
+                chunkDirNeutralSections.put(dirChunkLower, dirChunkUpper, incomingNeutralSection)
+
+                // TODO: remove this priority given to announcement.
```
(it's OK but useless) We have to remove this, too.

multun replied:
I suggest we do it in a separate PR, to ensure this one changes as little behavior as possible

Erashin commented on core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/new_impl/RawInfraFromRjsBuilder.kt:
```diff
@@ -0,0 +1,787 @@
+        // at this point, we have trackRangeIndex pointing to the track range which contains the
+        // entry
+        // detector
+        // we need iterate on track range until we reach the track range which contains the end
```
```suggestion
        // we need to iterate on the track range until we reach the track range containing the end
```

Erashin reviewed (approved):
I globally agree with @younesschrifi. This is a very good step (and an essential one, goodbye java infra) towards the right direction. Good job :)

eckter reviewed (approved):
LGTM, I skimmed through the first commit but I did review large parts of it

I tested on STDCM-related stuff, there's a new bug but it's not related to this specific PR (something with conditional speed limits)

I had some weird stuff with gradle / intellij at first when moving to this branch, something about broken cache files. But it fixed itself before I could figure out what was going on.

clarani reviewed (approved):
LGTM, tested :white_check_mark:
